device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/LoggingFlowController.java

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.sdk.iot.device.transport.amqps;
+
+import lombok.extern.slf4j.Slf4j;
+import org.apache.qpid.proton.engine.BaseHandler;
+import org.apache.qpid.proton.engine.Event;
+import org.apache.qpid.proton.engine.Link;
+import org.apache.qpid.proton.engine.Receiver;
+
+/**
+ * Copy of {@link org.apache.qpid.proton.reactor.FlowController} but with some small edits for logging purposes.
+ *
+ * Use this FlowController to automatically flow link credit back to the service from this client's receiver links after
+ * each delivery on the receiver link.
+ */
+@Slf4j
+public class LoggingFlowController extends BaseHandler
```

Comment:
Can we inherit from org.apache.qpid.proton.reactor.FlowController instead?

Reply:
The relevant method that is being overidden from FlowController is a private method, so no.